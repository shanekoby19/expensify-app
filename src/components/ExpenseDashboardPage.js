import React from 'react';
import ConnectedExpenseList from './ExpenseList';
import ConnectedExpenseListFilters from './ExpenseListFilters';

export default () => (
    <div>
        <ConnectedExpenseListFilters />
        <ConnectedExpenseList />
    </div>
);