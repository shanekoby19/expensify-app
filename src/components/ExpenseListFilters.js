import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }

    onTextChange = (e) =>  {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    }

    onSortChange = (e) =>  {
        this.props.setTextFilter(e.target.value);
    }

    render = () => {
        return (
            <div>
                <input 
                    type='text' 
                    value={this.props.filters.text} 
                    onChange={this.onSortChange}
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onTextChange}
                >
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId={this.props.filters.startDateId}
                    endDate={this.props.filters.endDate}
                    endDateId={this.props.filters.endDateId}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        ) 
    }
}

const mapStateToProps = (state) => ({
        filters: state.filters,
})

const mapDispatchToProps = (dispatch => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
}))

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);