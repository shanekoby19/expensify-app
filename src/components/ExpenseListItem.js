import React from 'react';
import { Link } from 'react-router-dom';

export default ({description, amount, createdAt, id, dispatch}) => (
    <div>
        <h2>
            <Link to={`/edit/${id}`}>{description}</Link>
        </h2>
        <p>{amount} - {createdAt.format('MMMM Mo')}</p>
    </div>
)