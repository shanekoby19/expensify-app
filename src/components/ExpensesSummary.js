import numeral from 'numeral';
import React from 'react';
import { connect } from 'react-redux';
import getAllExpenses from '../selectors/expenses';
import getTotalAmount from '../selectors/expenses-total'

export const ExpensesSummary = (props) => {
    const filteredExpenses = getAllExpenses(props.expenses, props.filters);
    return (
        <div>
            <p>Viewing {filteredExpenses.length} {filteredExpenses.length === 1 ? 'expense' : 'expenses'} totaling {numeral(getTotalAmount(filteredExpenses) / 100).format('$0,0.00')}.</p>
        </div>
    )
}

const mapStateToProps = (state) => ({
    expenses: state.expenses,
    filters: state.filters,
})

export default connect(mapStateToProps)(ExpensesSummary)