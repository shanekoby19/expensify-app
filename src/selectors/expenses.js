import moment from 'moment';

export default ( expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter(expense => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = text ? expense.description.toLowerCase().includes(text.toLowerCase()) : true;
        
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expenseOne, expenseTwo) => {
        if(sortBy === 'amount') {
            return expenseOne.amount >= expenseTwo.amount ? -1 : 1;
        }
        else if(sortBy === 'date') {
            return expenseOne.createdAt >= expenseTwo.createdAt ? -1 : 1;
        }
    })
}