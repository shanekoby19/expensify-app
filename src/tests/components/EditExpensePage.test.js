import React from 'react';
import { test, expect } from '@jest/globals';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage
            expense={expenses[0]}
            editExpense={editExpense}
            removeExpense={removeExpense}
            history={history}
        />
    )
})

test('should render the EditExpensePage component correctly', () => {
    expect(wrapper).toMatchSnapshot();
})

test('should dispatch edit expense to change an expense.', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should dispatch the remove expense action to remove an express', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[0].id });
});