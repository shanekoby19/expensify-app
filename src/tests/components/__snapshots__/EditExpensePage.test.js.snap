// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the EditExpensePage component correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    amount={195}
    createdAt={"1970-01-01T00:00:00.000Z"}
    description="Gum"
    id="1"
    note=""
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
